Core: Generic Entrance Rando

BadMagic100 opened the pull request:
## What is this fixing or adding?

This PR implements a generic entrance rando approach which can be run by worlds before item fill. Features include:

* Shuffling with performant grouping, e.g. "only match right transitions with left transitions", "only shuffle transitions in some area of the map with each other" and so on, including ordered or unordered 1-to-many mapping
* Support for custom constraints by deriving the Entrance class, both by determining if a given exit is visible to randomization and by determining whether a proposed entrance placement is valid
* Client-friendly output of randomization results
* Coupled (A->B implies B->A) and uncoupled (A->B does not imply B->A) randomization modes

## How was this tested?

Unit tests - ran all unit tests, including adding new ones to cover the new functionality in generic ER. Some helpers for setting up (multi)worlds for tests were moved from fill tests to `__init__.py` since there was utility here as well and could reasonably be needed elsewhere in the future
Practical tests - Messenger, KDL3, and minit all have working or WIP implementations which have been able to generate as expected

I have not tested generation for games without an ER implementation (i.e., worlds which don't have unit tests), but this is mostly additions so I don't believe that is really needed; the only actual changed behavior is that CollectionState can now conditionally (default off) handle entrances without a connected_region.

## Related PRs (non-blocking)

#2083 will impact the tests here. 
#2427 would improve performance; currently we use all_state which will slow down with multiworld size, but we only need a single world's state which could be provided by that.

alwaysintreble reviewed (commented):
`EntranceRando.py` should be renamed to `entrance_rando.py` since we're trying to move to modern PEP8 conventions. code lgtm. i've looked over it a ton and gave it a review when it was still in its early stages since i use it for my ER and all issues i had have been resolved.

BadMagic100 commented on test/general/__init__.py:
```diff
@@ -27,3 +28,51 @@ def setup_solo_multiworld(world_type: Type[World], steps: Tuple[str, ...] = gen_
+        multiworld.game[player_id] = f"Game {player_id}"
+        multiworld.worlds[player_id] = world
+        multiworld.player_name[player_id] = "Test Player " + str(player_id)
+        region = Region("Menu", player_id, multiworld, "Menu Region Hint")
```
I cut and pasted this from the fill tests since it became useful for ER tests. so I didn't want to mess with it

BadMagic100 commented on BaseClasses.py:
```diff
@@ -657,6 +659,7 @@ def __init__(self, parent: MultiWorld):
         self.path = {}
         self.locations_checked = set()
         self.stale = {player: True for player in parent.get_all_ids()}
+        self.allow_partial_entrances = allow_partial_entrances
```
Probably, the simple reason it isn't is because I didn't think about or test state.copy with this so will add

BadMagic100 commented on test/general/__init__.py:
```diff
@@ -27,3 +28,51 @@ def setup_solo_multiworld(world_type: Type[World], steps: Tuple[str, ...] = gen_
     return multiworld
+
+
+def generate_multiworld(players: int = 1) -> MultiWorld:
```
Might just docstring it tbh, the whole point of moving it from test_fill was other tests wanted it so docs are probably appropriate in case even more want it in the future

Jouramie reviewed (approved):
I have read all of it, but I can't say I understood any of it... 

I am however eager to use this with Stardew, and it will be much easier to understand when diving in. As this is somewhat in isolation at the moment (no doubt some have tested the integration already and PRs will open pretty fast once it's merged), I would not be afraid of merging this. Finding issues will be much easier once it is in main than in a PR.

Also, feel free to dismiss my nitpicking.

BadMagic100 commented on EntranceRando.py:
```diff
@@ -0,0 +1,414 @@
+            self._lookup.setdefault(entrance.er_group, []).append(entrance)
+
+        def remove(self, entrance: Entrance) -> None:
+            group = self._lookup.get(entrance.er_group, [])
```
Yeah it would be though the amount of time I spent checking probably indicates this implementation is more clear and definitely does what it should rather than being reliant on the implementation of getitem

BadMagic100 commented on BaseClasses.py:
```diff
@@ -803,25 +808,36 @@ def remove(self, item: Item):
         self.parent_region = parent
         self.player = player
+        self.er_group = er_group
+        self.er_type = er_type
```
I did think about it but was weighing against being clear these fields (which are optional) are only relevant to ER and I didn't really want to spell out entrance_rando_type. But this is a good point, I will think about it some more

BadMagic100 replied:
I've renamed these to `randomization_group` and `randomization_type` - this makes it clear they are only for randomization, including "entrance" in the name is redundant because they are members of entrances.

BadMagic100 commented:
@Jouramie would be interested where you felt the gaps were, as it might be addressable by better docs especially if it's something user facing. Admittedly I probably ought to write up an actual explainer with examples, visuals, and faq but I wanted to get the PR out into the world so I haven't written it

alwaysintreble commented:
i definitely think the docstrings need to be trimmed down to only the bare minimum and everything else moved into a separate doc with examples etc., but i don't think it's really a blocker atm and would be fine to be done at a later date.

NewSoupVi commented:
I agree that an actual doc with example uses and stuff would be really really good for this instead of having it all in the docstrings

Jouramie commented:
I do think having examples will be a good way to understand how the generic ER works. Right now, I have no doubt I could use it in Stardew from the docstring + trial and error, but it would be easier to see how others have integrated it. 

I feed like the user facing API is well documented. It's more the internal of the algorithm that is hard to understand without going deep into it, maybe with a debugger.

BadMagic100 commented:
Regarding docstrings length, I think the docstrings are actually at just the right length, there are quite a lot of prerequisites to meet and they should be documented in the API ref. Honestly I would even venture to say it is quite short compared to some other docstring/javadoc/xmldoc I have seen especially if we plan to generate API reference from docstrings (which I understand is in progress currently), which tends to drive more verbose in-code documentation. I will add a doc with actual examples and visuals in my revision, a ton of discussion from the early design and implementations should probably be collated there and it seems like it is much needed

Regarding understanding the internals, I am not sure what to do about it but it does concern me a little bit, ideally someone other than me should be able to figure out what is going on reasonably easily. I regard this as similar complexity to fill which I was able to sit down and figure out within an hour or so. Though I also had a plain english explanation of how it should work, and from that perspective the above doc article may also be helpful

alwaysintreble commented:
> Honestly I would even venture to say it is quite short compared to some other docstring/javadoc/xmldoc I have seen especially if we plan to generate API reference from docstrings (which I understand is in progress currently), which tends to drive more verbose in-code documentation.

what i have so far is at #2909 and there's a thread in discord as well. plan is to add everything from our docs folder to it also, so i think it makes more sense to have explanations, examples, or detailed descriptions in docs instead of in the docstrings. i don't think the docstrings being long is necessarily an issue but i do think they should only give a brief overview of what the method does and requirements of the params etc.

adding a doc that explains how the fill methods actually work is probably a good idea too.

BadMagic100 commented:
> i do think they should only give a brief overview of what the method does and requirements of the params etc.

Right, I agree but would argue that the current state is just about as brief as can be to summarize the expected inputs and behavior of the function. It's a complex function. There is certainly some stuff (clarifications such as the requirements for coupling to work) that could be trimmed with a separate doc but order of magnitude I expect it should end roughly the same size

qwint reviewed (approved):
ran tests with 3.8/3.11
weighed in during development of the PR
am utilizing the branch for developing entrance rando for Minit
and re-reviewed with a fresh set of eyes recently

I did not fully review:
* the entrance/group lookup code
* the test changes past a short comparison with the tests being moved 
* have not tested uncoupled at all, but did review the code for it

BadMagic100 commented:
An observation - grubhunt goal seems to make playthrough very slow. with 5 hks itemlinking grubs + linked replacements playthrough took the better part of a 9min generation

BadMagic100 commented:
Quick link for being able to see the diagrams as diagrams (rather than markdown): https://github.com/BadMagic100/Archipelago/blob/generic-entrance-rando/docs/entrance%20randomization.md

BadMagic100 commented on entrance_rando.py:
```diff
@@ -0,0 +1,414 @@
+    # place the menu region and connected start region(s)
+    er_state.collection_state.update_reachable_regions(world.player)
+
+    # stage 1 - try to place all the non-dead-end entrances
+    while entrance_lookup.others:
+        if not find_pairing(False, True):
+            break
+    # stage 2 - try to place all the dead-end entrances
+    while entrance_lookup.dead_ends:
+        if not find_pairing(True, True):
+            break
+    # stage 3 - connect any dangling entrances that remain
+    while entrance_lookup.others:
+        find_pairing(False, False)
+    # stage 4 - last chance for dead ends
+    while entrance_lookup.dead_ends:
+        find_pairing(True, False)
```
Good idea

korydondzila commented on test/general/test_er.py:
Can you rename this file to `test_entrance_rando` to match the logic file?

Jouramie reviewed (approved):
The doc article is really helpful to understand how the algorithm works. I renew my approval.

ScipioWright commented on entrance_rando.py:
```diff
@@ -345,18 +345,18 @@ def find_pairing(dead_end: bool, require_new_regions: bool) -> bool:
     # stage 4 - last chance for dead ends
     while entrance_lookup.dead_ends:
-        find_pairing(True, False)
+        find_pairing(dead_end=True, require_new_regions=False)
```
Stage 3 is to connect any dangling entrances, and stage 4 is the "last chance for dead ends". What's the consideration here?
I'd assume that any dead ends would be connected in stage 2, and that after stage 3 there should be no available entrances left, unless someone marked an area as a dead end that isn't actually a dead end.

ScipioWright replied:
So, why don't stage 1 and 2 guarantee access to all regions? Ideally, those event-gated ones would not stop it from pairing all of the scenes. If the latch/switch/whatever you wanna call it is in a dead end, that dead end shouldn't be called a dead end so that it can get paired in the first go-around.

korydondzila replied:
I am curious about this, in particular the handling of dead ends in how the map itself is generated. Granted I don't personally understand all the inner workings of AP, and haven't fully read the logic, but from what does seem clear to me and coming from a graph generation perspective I do have some thoughts.

The first is, why not pair dead ends (regions with a single entrance/exit) with regions that are not dead ends first? The reason being that given the list of regions that have only 1 entrance (or those with entrances but no exits), if you always pair one with another region that is not a dead end then all your dead ends should get paired.

The second being, assuming you do the first, the dead end list can update with regions that were other (not dead ends) but can now be considered dead ends, for the purpose of generation. That is to say if you have room X which is a dead end, and room Y which is not dead end but only has 2 entrance exits, once you pair X to one of the entrances with Y, Y can now be thought of as a dead end (it's a region that is now itself and X).

If you follow this logic, you should end up in a situation where, after you process all the dead ends (and regions that "become" dead ends), you only have others which you can hook up as normal (making them dead ends where possible). Or you have no regions that are not dead ends, in which case there should only be 2 dead ends which could then be connected (I'd need to check this, but if more then you'd just connect as normal w/o creating orphans). This should also guarantee a connected graph.

There are of course multiple ways to do generation, the general logic is similar, but does imply 3 states not 4. You have dead ends and others, you have only others, or you just have dead ends.

(The require new regions might muddle this, not sure, but it seems like that would create less variety?)

Also feel free to completely ignore my thoughts if it doesn't fit or ask questions if I don't make sense.

BadMagic100 replied:
> The first is, why not pair dead ends (regions with a single entrance/exit) with regions that are not dead ends first? The reason being that given the list of regions that have only 1 entrance (or those with entrances but no exits), if you always pair one with another region that is not a dead end then all your dead ends should get paired.

The big reason this doesn't work is because of Entrance access logic which (as outlined in the documentation) is used to ensure to ensure a graph is generated which will not make the chance of fill errors notably higher than in a non-ER context. For this to work you would either have to limit the pairings only to exits which require only randomized items (potentially extremely difficult to check) or just ignore logic and roll the dice and hope that no transition ends of being directly or transitively self-locking

BadMagic100 replied:
The rest I think will be best if I sit and reread my code as it has now been months since I wrote it, and I can't remember all the details of implementation.

BadMagic100 replied:
Ah, here's the consideration for stage 4. Stages 1 and 2 get to pass if every remaining target points to an already-placed region. So if there is a dead end region with 2 one-way entrances but no exits, for example, that would need stage 4 to connect the second edge in.

In an ideal world stages 3 and 4 do not give any preference order to dead ends vs non dead ends, as both are serving the same purpose of trying to interconnect already-available regions. Maybe instead of dead-end being boolean it should be a flag enum going into group lookup to achieve that but I think the effect is negligible because this edge case is probably so rare and the current setup allows a good deal of code reuse.

BadMagic100 replied:
Actually this made me realize an issue with the logic for determining deadendedness, namely that a 2-way into/out of a region is a dead end in coupled but is not a dead end in uncoupled

korydondzila replied:
I didn't even pay attention to that, but yeah I knew exactly which part you were talking about when you just mentioned it. [this below](https://github.com/ArchipelagoMW/Archipelago/pull/2883#discussion_r1618221868)

I do think that in uncoupled mode dead ends would be those regions that have an entrance but the exit has already been used. I think this is partly why it was odd to me that dead ends weren't tracked beyond the initial entrance lookup creation. (whether this actually matters for this generation is unknown to me)

BadMagic100 replied:
The definition of a dead end is a target which cannot provide access to new randomized exits even when ignoring all logic conditions (this is also specified in the documentation). So you never need or want to recompute it. Here are some illustrative cases which I believe are not worth putting in the docs

```mermaid
graph LR
  x1:::hidden-->|e1| A
  x2:::hidden-->|e2| A
  classDef hidden display:none;
```
Both entrances to A (e1 and e2) are always dead ends because there is no way out, and therefore placing it may cap off the region graph mid generation (this is why we care about dead ends, we want to make sure that we don't close all our routes to expand too early).

---

```mermaid
graph LR
  x1:::hidden<-->|e1| A
  classDef hidden display:none;
```
The entrance to A (e1) is a dead end in uncoupled mode, because once you get in, you can go back out and end up elsewhere. It is a dead end in coupled mode as the reverse exit (also called e1 because this is the only way we have to do coupling) gets auto-placed to a region you have been already.

---

```mermaid
graph LR
  x1:::hidden-->|e1| A -->|e2| x2:::hidden
  classDef hidden display:none;
```
The entrance to A (e1) is never a dead end because it always can give access to the exit of A (e2)

BadMagic100 replied:
I ended up adjusting the comments and docs around this as well in the latest commit, when I switched the order to prevent ER from hardlocking

BadMagic100 commented on entrance_rando.py:
```diff
@@ -0,0 +1,366 @@
+
+            for exit_ in region.exits:
+                # randomizable and not the reverse of the start entrance
+                if not exit_.connected_region and exit_.name != entrance.name:
```
`exit_.name != entrance.name` is only relevant in coupled mode

BadMagic100 reviewed (commented):
Today I learned you cannot request changes on your own PR

alwaysintreble commented on entrance_rando.py:
```diff
@@ -0,0 +1,366 @@
+    # similar to fill, skip validity checks on entrances if the game is beatable on minimal accessibility
+    perform_validity_check = True
+
+    def do_placement(source_exit: Entrance, target_entrance: Entrance):
```
```suggestion
    def do_placement(source_exit: Entrance, target_entrance: Entrance) -> None:
```

alwaysintreble reviewed (commented):
code all lgtm and i've tested it quite a bit with messenger. all comments i have are just docs related. the only question i really have is why `get_target_groups` is a callable. since it only passes the string i can't really think of anything it can do that a dictionary wouldn't be sufficient for. even the provided examples could just be built into a dictionary

BadMagic100 commented on docs/entrance randomization.md:
```diff
@@ -0,0 +1,377 @@
+
+Some important terminology to understand when reading this doc and working with ER is listed below.
+
+* Entrance rando - sometimes called "room rando," "transition rando," "door rando," or similar,
```
My understanding was that it is correct grammar to have it inside, apparently that is not true however a quick google for "does comma go inside or outside quotes" brings back plenty of reputable sources recommending inside, at least in American English. Here is a secondary source with this exact example, even: https://ell.stackexchange.com/questions/257851/should-the-comma-be-inside-or-outside-of-the-quotation-marks

I guess this raises the question though, is there an actual preferred style guide for docs? If not there should be (the opportunity to use AP style is very funny to me).

black-sliver replied:
I think the comma is inside the quotes only if it is used for a quote - i.e. someone saying something. Here the quotes are used to mark an expression. I could be wrong as non-native speaker.

(I would go for `"room rando", ...`)

black-sliver replied:
After googling for a bit, I came to the conclusion that it depends on AmE vs. BrE:
https://www.reddit.com/r/grammar/comments/11c3aeg/when_using_scare_quotes_does_punctuation_go/

While this breaks my programmer brain, I think it would be correct since we mostly use AmE.

black-sliver replied:
This is kind of just another source for what you found, but I wanted to clarify that I came to the same conclusion, and that I think we'll go with AmE style until decided otherwise.

BadMagic100 commented on docs/entrance randomization.md:
```diff
@@ -0,0 +1,377 @@
+   and create your events before you call ER if you want to guarantee a correct output.
+
+If the conditions above are met, you could theoretically do ER as early as `create_regions`. However, plando is also 
+a consideration. Since item plando happens before `pre_fill` and modifies the item pool, doing ER in `pre_fill` is the
```
I wrote it with an implied "immediately" in mind but sure, can update accordingly

BadMagic100 commented on docs/entrance randomization.md:
```diff
@@ -0,0 +1,377 @@
+If the conditions above are met, you could theoretically do ER as early as `create_regions`. However, plando is also 
+a consideration. Since item plando happens before `pre_fill` and modifies the item pool, doing ER in `pre_fill` is the
+only way to account for placements made by item plando, otherwise you risk impossible seeds or generation failures. 
+Obviously, you may also want to perform ER after entrance plando as well.
```
How so? Throughout the doc (including in this line even) it is specifically called out that "you,"  ie, the reader, are responsible for performing ER/calling `randomize_entrances`, I guess I don't see what the proposed change is

BadMagic100 replied:
Ah, I think I misunderstood the first time, updating to be more clear that worlds would handle their own entrance plando (and if they do, they likely want to do it before ER)

BadMagic100 commented:
> why get_target_groups is a callable

As discussed in discord, made this a dictionary of any hashable object (specifically allowing int groups) to improve performance. We also now no longer check/append the default group and allow failed gen instead. There is instead a convenience function which applies a callable to all observed groups in the world to pre-build the lookup

alwaysintreble reviewed (approved):
very minor nitpicks but my concerns were appropriately addressed

BadMagic100 commented on BaseClasses.py:
```diff
@@ -9,7 +9,7 @@
 from argparse import Namespace
 from collections import Counter, deque
-from collections.abc import Collection, MutableSequence
+from collections.abc import Collection, MutableSequence, Hashable
```
Correct, must have missed it in the most recent commit. Hard to catch which linter errors are real when half the file is yellow

NewSoupVi commented:
Hey! I'm going to be looking at this PR in the coming days.

I had a question about one way and two way entrances. Here's a condensed/cleaned up excerpt from a very chaotic [Discord conversation](https://discord.com/channels/731205301247803413/1159521365024591985/1273378645624361023) that should explain it:

**NewSoupVi**:
I just read the docs and saw this:
> To ensure that all transitions can be accessed in the game, one-ways are only randomized with other one-ways and two-ways are only randomized with other two-ways.
Interesting! I believe OoTR (like normal, testrunner OoTR) does not do this in certain situations 
You can go through a loading zone that is two-way in vanilla, and be surprised by a one-way exit you're unable to go back through 
As I understand it, this function is intended to be overridable though, so if a world wants to do that, it can
If this is not a supported use case that seems like a slight oversight to me, although probably not relevant for most devs

**qwint**:
connect something with an entrance but no exit to a two way, what happens when they enter the two way side?

**NewSoupVi**:
Ok so that one is actually fair
So I guess I'd be ok with Two Way Entrances taking you to One Way Exits (perfectly well defined, an exit just becomes permanently unusable), but not One Way Entrances taking you to Two Way Exits (undefined what should happen when the two way entrance is taken on the other side)
Does this make sense?

**qwint**:
yes

**NewSoupVi**:
In that case, is it something you can do with the current generic ER implementation? It's not at all a dealbreaker if not, this is a specific use case that most devs won't care about and they'll just be happy generic ER exists for them to make use of. But it would be nice

NewSoupVi commented:
I now understand why this cannot work thanks to an explanation in Discord by BadMagic.

NewSoupVi commented on entrance_rando.py:
```diff
@@ -0,0 +1,388 @@
+                    and world.multiworld.has_beaten_game(er_state.collection_state, world.player):
+                # ensure that we have enough locations to place our progression
+                accessible_location_count = 0
+                prog_item_count = len(er_state.collection_state.prog_items[world.player])
```
`CollectionState.prog_items` is a **Dict** (specifically, Counter) per player.

This means that someone's `prog_items` look like this:

```
{
    "Sword": 3,
    "Shield": 1,
}
```

`len(prog_items[player])` would now return `2`, when there are `4` items collected.

This seems unintentional. It if isn't intentional, you'll want

```
sum(er_state.collection_state.prog_items[world.player].values()
```

NewSoupVi commented on entrance_rando.py:
```diff
@@ -0,0 +1,388 @@
+            # if we're on minimal accessibility and can guarantee the game is beatable,
+            # we can prevent a failure by bypassing future validity checks. this check may be
+            # expensive; fortunately we only have to do it once
+            if perform_validity_check and world.options.accessibility == Accessibility.option_minimal \
```
We discussed this on the PR, but this indicates that `minimal` still tries to make a complete map, and only "prevents a failure state" that would normally crash.

As a player, I would expect that `minimal` + ER would lead to half my regions being inaccessible most of the time.

However, as I said in that discussion, I also don't think there's a super elegant solution to this. We came up with some things that could be tried, but for me, they all had shortcomings.

Either way, it's something I find very much "noteworthy"

BadMagic100 replied:
Updated docs.

NewSoupVi commented on entrance_rando.py:
```diff
@@ -0,0 +1,388 @@
+                f"All unplaced entrances: {unplaced_entrances}\n"
+                f"All unplaced exits: {unplaced_exits}")
+
+    er_targets = [entrance for region in world.multiworld.get_regions(world.player)
+                  for entrance in region.entrances if not entrance.parent_region]
+    exits = [ex for region in world.multiworld.get_regions(world.player)
```
It *might* be nice if you could optionally provide these yourself to do GER in "stages", but I also realize there's some problems with this approach since GER fundamentally happens from the Entry region in a front-fill way, and the "groups" feature seems to mostly take care of anything you'd want to do with "separate randomisation" anyway. Anything that doesn't fit it the world would probably just do itself, I suppose

BadMagic100 replied:
It actually used to be pass-it-yourself but every implementation ended up doing this so we  agreed it would be better to just take the work out. There are a few non-hacky ways the world could do it in stages, the main ones being to iteratively define the entrances and call GER repeatedly, or to use is_valid_source_transition and can_connect_to to filter out later stages until all of the earlier stages are complete. I suspect the vast majority of users will be quite happy with it as is though so treating such a use case as an edge case seems fine to me

NewSoupVi replied:
Hmmm but you could still let a world optionally provide these, right?
Like how `sweep_for_events` works:

https://github.com/ArchipelagoMW/Archipelago/blob/main/BaseClasses.py#L758-L760

Idk I don't think it's a huge deal, but it seems somewhat simple to add this in this case

NewSoupVi commented on entrance_rando.py:
```diff
@@ -0,0 +1,388 @@
+                  for entrance in region.entrances if not entrance.parent_region]
+    exits = [ex for region in world.multiworld.get_regions(world.player)
+             for ex in region.exits if not ex.connected_region]
+    if len(er_targets) != len(exits):
+        raise EntranceRandomizationError(f"Unable to randomize entrances due to a mismatched count of "
```
Oh this is interesting

I feel like "I'm gonna make a bunch of extra entries that don't need to be connected to" could be a valid use case, e.g. when "randomizing warp song destinations" where you want the 6 warp songs to go to completely random regions

Again, could just be something the world would do itself though.

BadMagic100 replied:
Yeah that doesn't seem like a GER problem to me. Having a matched number of "somewhere to go from" and "somewhere to go" seems pretty fundamentally important on a conceptual level, in addition to a technical one

ScipioWright replied:
In a case like that, the world dev could choose which of those 6 random regions to use in advance

NewSoupVi replied:
I'll be solving this problem with The Witness soon, maybe it's something you're interested in:

I'll basically have a directed region graph where not everything can be reached from the starting region, and I want to be able to pick warps (connections from Entry to random regions) **such that** everything is reachable.

I'm ok with it not being a GER problem, but it's an ER problen and a fun graphing problem, so we'll see how I'll solve it :)

Anyway, consider this resolved if you want to

BadMagic100 replied:
Yeah within the scope of GER, for a case like this, what I would immediately think to do is assign groups for the "islands" (possibly psuedo-randomly if desired) and have ER do grouping within the island subgroups. Then you can pre-place warps from start to "somewhere in each island" and GER would solve the rest

NewSoupVi reviewed (commented):
The code is actually quite nice. I had one genuine concern and a couple of comments. Some of my comments are [on Discord](https://discord.com/channels/731205301247803413/1159521365024591985/1273422442123231242) because github actually went down during this review, but they were all addressed. I did choose to bring one of them over to this review, but yeah.

I wanna test more, and I haven't reviewed `test_entrance_rando.py` yet
Apart from the prog_items thing, this can be seen as a "code approval" of BaseClasses.py and entrance_rando.py, though.

NewSoupVi commented on test/general/test_entrance_rando.py:
```diff
@@ -0,0 +1,376 @@
+        multiworld = generate_test_multiworld()
+        generate_disconnected_region_grid(multiworld, 5)
+
+        lookup = EntranceLookup(multiworld.per_slot_randoms[1], coupled=True)
```
Deprecated, may be hard-deprecated eventually. Use `multiworld.worlds[1].random`

BadMagic100 replied:
To my credit, at the time these were originally written, some peculiarities of world loading prevented this from working because world.random was not getting initialized in tests, fixing that was considered out of scope for this PR.

NewSoupVi commented on test/general/test_entrance_rando.py:
```diff
@@ -0,0 +1,376 @@
+        retrieved_targets = lookup.get_targets([ERTestGroups.TOP, ERTestGroups.BOTTOM],
+                                               False, False)
+        prev = None
+        group_order = [prev := group.randomization_group for group in retrieved_targets if prev != group.randomization_group]
```
Line length violation

Would let this go if it was world code, but new core code should follow our style guide

NewSoupVi commented on test/general/test_entrance_rando.py:
```diff
@@ -0,0 +1,376 @@
+
+
+    def test_ordered_targets(self):
+        """tests that get_targets does not shuffle targets between groups when requested"""
+        multiworld = generate_test_multiworld()
+        generate_disconnected_region_grid(multiworld, 5)
+
+        lookup = EntranceLookup(multiworld.per_slot_randoms[1], coupled=True)
+        er_targets = [entrance for region in multiworld.get_regions(1)
+                      for entrance in region.entrances if not entrance.parent_region]
+        for entrance in er_targets:
+            lookup.add(entrance)
```
This duplicated code could be in `setUp`, no?

BadMagic100 replied:
My opinion/experience with moving this kind of setup to a before-each-test setup method is that
1. It makes it difficult to follow what state is being tested, as your "arrange" is nowhere near your "act" and "assert" in source
2. As soon as you try to generalize the setup you will find many cases where it doesn't work well and you have to replicate the whole setup anyway

Since the duplicated block is small I believe duplicating it is cleaner, especially in a test scenario (in production code it might be different)

NewSoupVi commented on test/general/test_entrance_rando.py:
```diff
@@ -0,0 +1,376 @@
+        generate_disconnected_region_grid(multiworld, 5)
+        seen_placement_count = 0
+
+        def verify_coupled(state: ERPlacementState, placed_entrances: List[Entrance]):
```
Could use `_: ERPlacementState` to signal that the variable is unused. My IDE is happier when I do this, anyway

NewSoupVi reviewed (changes requested):
Unit tests also look good to me with one important change & some minor nitpicks

BadMagic100 commented:
Addressed various review comments/discord discussions in the most recent commit. It would be nice to have unit tests added for the 3 major behavior changes (speculative sweep, sweep for events, prog item count)

Summary of changes:
* Updated docs on various points, including the definition of one-ways and documenting the behavior on minimal.
* Added a missing sweep_for_events
* Correct how progression items are counted for minimal
* Cleanup of tests
* Implemented a speculative sweep in stage 1 which will help ensure that there are functionally-accessible exits after the last connection is made (covering for region/event-based rules on exits more carefully)
* Added `dead_end: bool` parameter to `Entrance.can_connect_to` as there are a variety of use cases where an implementer may be interested to know whether the current stage is placing dead ends (a trivial example would be "I need to ensure that no paths get too long so after some point only dead-ends are valid placements")

NewSoupVi commented:
This PR is affected by the change from `sweep_for_events` to `sweep_for_advancements` and will fail unit tests when main is merged into it

Rather simple change though :)

Ishigh1 commented:
Just finished implementing ER on Oracle of Seasons using this. 
Overall, works kinda simply, it's a bit unintuitive that if you want to make an existing transition randomized, saving it somewhere in the world to then disconnect it later was the easiest way I've seen. I'd have expected GER to disconnect some entrances by itself
Performance wise, it seemed good, at least not absurdly slow
Failure rate is where it hurts more, as I talked about on discord, there's some structures that seem to highly increase failure rates; the algorithm sometimes put aside too many dead ends and ends up with too many and fails

BadMagic100 commented:
Current state of development - I have added fixes for all of the feedback provided here except the one just above, about the dead ends. However, adding tests for these fixes revealed that in fact it just shoves the error off to a different place. Some rewriting will be required to get the tests to pass; once they do we should be good to go again.

beauxq commented on entrance_rando.py:
```diff
@@ -0,0 +1,429 @@
+        def __iter__(self):
+            return itertools.chain.from_iterable(self._lookup.values())
+
+        def __str__(self):
+            return str(self._lookup)
+
+        def __repr__(self):
+            return self.__str__()
```
```suggestion
        def __repr__(self):
            return str(self._lookup)
```
`__str__` automatically defaults to `__repr__`
If you just define `__repr__` then `__str__` will automatically be the same thing.

BadMagic100 commented:
Today's update should cover both remaining concerns
1. The issue where too many exits are consumed to place dead ends should be resolved by flipping the order of stages 3 and 4 (renamed to stage 3a and 3b now to better capture their purpose based on a much earlier discussion)
2. The issues with logic dependent on other regions is resolved by treating any region which could be needed by logic (has a placed progression item, is used in an indirect condition, or has randomized exits as was previously the case) as a non-dead-end. This should be very safe because the speculative sweep step is still present to prevent the graph from being closed off prematurely.
3. Updated docs according to both of these changes.

NewSoupVi commented on docs/entrance randomization.md:
```diff
@@ -72,6 +75,11 @@ logic to generate a valid world layout by connecting the partially connected edg
 that, your region graph might look something like the following diagram. Note how each randomizable entrance/exit pair
 (represented as a bidirectional arrow) is disconnected on one end.
 
+> [!WARNING]
+> It is highly recommended to use explicit indirect conditions when using Generic ER. Failure to do so may lead to an
+> increased failure rate if your dead end regions are required in logic rules as Generic ER would have no way to
+> determine they are required when performing randomization and treat them accordingly.
```
Wondering if this should straight up error, haha

NewSoupVi replied:
I guess there may be worlds that are super open and have like, a single indirect condition in a non dead end region. And it never leads to problems. ~~But also a world like that should be using explicit indirect conditions~~

BadMagic100 replied:
I think you are probably right, I can make this error. The better option imo is to make this not a warning ie to solve for implicit indirect conditions but I'm pretty sure that is not possible without some change to rules that allows better static analysis

NewSoupVi replied:
It's kind of possible to detect if you monkeypatch the hell out of core, but yeah, I wouldn't recommend it

NewSoupVi commented:
Also, I hate to do this since this PR has been open *forever*, but we probably shouldn't merge anything that uses old typing anymore. Hopefully this is at least an easy fix? Ruff for example can do this automatically

NewSoupVi commented:
I think other than that, this is good to go - so hopefully that prospect can be the motivation to make those final changes haha
